// earl@1.1.0/dist/cjs/validators/other/toExactlyEqual.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.toExactlyEqual = void 0;
const expect_js_1 = require("../../expect.js");
const index_js_1 = require("../../format/index.js");
(0, expect_js_1.registerValidator)('toExactlyEqual', toExactlyEqual);
function toExactlyEqual(control, expected) {
    const actualInline = (0, index_js_1.formatCompact)(control.actual);
    const expectedInline = (0, index_js_1.formatCompact)(expected);
    control.assert({
        success: sameValueZero(control.actual, expected),
        reason: `The value ${actualInline} is not the exact same value as ${expectedInline}, but it was expected to be.`,
        negatedReason: `The value ${actualInline} is the exact same value as ${expectedInline}, but it was expected not to be.`,
        actual: control.actual,
        expected,
    });
}
exports.toExactlyEqual = toExactlyEqual;
function sameValueZero(x, y) {
    if (typeof x === 'number' && typeof y === 'number') {
        // x and y are equal (may be -0 and 0) or they are both NaN
        return x === y || (x !== x && y !== y);
    }
    return x === y;
}
