// earl@1.1.0/dist/esm/Control.js
import { AssertionError } from './errors/index.js';
import { format } from './format/index.js';
export class Control {
    constructor(options) {
        var _a;
        this.isNegated = false;
        this.assert = (result) => {
            if (this.isNegated === result.success) {
                throw new AssertionError({
                    message: result.success ? result.negatedReason : result.reason,
                    stack: this._location.stack,
                    ...formatActualAndExpected(result),
                });
            }
        };
        this.fail = (result) => {
            throw new AssertionError({
                message: result.reason,
                stack: this._location.stack,
                ...formatActualAndExpected(result),
            });
        };
        this.name = options.name;
        this.actual = options.actual;
        this.isNegated = (_a = options.isNegated) !== null && _a !== void 0 ? _a : false;
        this._location = AssertionError.getLocation(this.name);
    }
    get file() {
        return this._location.file;
    }
}
function formatActualAndExpected(result) {
    if (!('actual' in result && 'expected' in result)) {
        return {};
    }
    if (typeof result.actual === 'string' &&
        typeof result.expected === 'string') {
        return { actual: result.actual, expected: result.expected };
    }
    return {
        actual: format(result.actual, null),
        expected: format(result.expected, result.actual),
    };
}
