// earl@1.1.0/dist/cjs/validators/objects/toHaveLength.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.toHaveLength = void 0;
const expect_js_1 = require("../../expect.js");
const index_js_1 = require("../../format/index.js");
const length_js_1 = require("../../matchers/objects/length.js");
(0, expect_js_1.registerValidator)('toHaveLength', toHaveLength);
function toHaveLength(control, expected) {
    const actualInline = (0, index_js_1.formatCompact)(control.actual);
    const expectedInline = (0, index_js_1.formatCompact)(expected);
    control.assert({
        success: (0, length_js_1.length)(expected)(control.actual),
        reason: `The value ${actualInline} does not have length ${expectedInline}, but it was expected to.`,
        negatedReason: `The value ${actualInline} has length ${expectedInline}, but it was expected not to.`,
        actual: getLength(control.actual),
        expected: expected,
    });
}
exports.toHaveLength = toHaveLength;
function getLength(value) {
    try {
        return value.length;
    }
    catch {
        return undefined;
    }
}
